### PR TITLE
Unprepare volume with Salt

### LINKED ISSUE #1475: Unprepare volume with Salt
**Component**: salt, storage

**Dependencies**:
- [x] #1473 
- [x] #1474

**Why this is needed**: To be able to remove device previously created with Salt

**What should be done**: Create a salt formula which will do the reverse of #1474

**Implementation proposal** (strongly recommended):

**Test plan**: We need to be able to completely "unprepare" a volume.

### PR DESCRIPTION
**Pull request by slaperche-scality**
**Component**:

salt

**Context**:

Cleanup resources when a `Volume` is deleted.

**Summary**:

Add a new Salt state to cleanup volumes.

**Acceptance criteria**: 

When you delete a `SparseLoopDevice`:
- the loop device disappeared from the output of `losetup -l`
- the backing file was removed from `/var/lib/metalk8s/storage/sparse/`

When you delete a `rawBlockDevice`:
- the disk is still formatted as before, data are left untouched

---

Closes: #1475

### CODE REVIEW

File: salt/_modules/metalk8s_volumes.py [hunk 1]
```diff
@@ -260,6 +307,29 @@ def format(self, force=False):
...
+        try:
+            device_path = '/dev/disk/by-uuid/{}'.format(
+                self.get('metadata.uid')
+            )
+            command = ' '.join(['losetup', '--detach', device_path])
+            _run_cmd(command)
+        except CommandExecutionError as exn:
+            # Gracefully handle if the loop device is already detached.
+            # Return code is too generic, have to check the error message…
+            if 'No such file or directory' not in exn.message:
```
Comment (slaperche-scality) [inline salt/_modules/metalk8s_volumes.py]: Not a big fan of this…

Maybe I should use `os.open` + `ioctl(fd, LOOP_CLR_FD, 0)`, error handling would be more fine-grained.

File: salt/_modules/metalk8s_volumes.py [hunk 2]
```diff
@@ -85,7 +87,7 @@ def provision(name):
         name (str): volume name
 
     Returns:
-        str: path to the associated loop device
+        None
```
Comment (NicolasT) [inline salt/_modules/metalk8s_volumes.py]: Then I guess this `Returns: None` can be removed altogether?

File: salt/_modules/metalk8s_volumes.py [hunk 3]
```diff
@@ -132,6 +134,42 @@ def format(name):
...
+        name (str): volume name
+
+    Returns:
+        bool: True if the backing storage device is cleaned up, otherwise False
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.is_cleaned_up example-volume
```
Comment (NicolasT) [inline salt/_modules/metalk8s_volumes.py]: You won't want to run this on `*`, no? A volume sits on a particular minion/host.

File: salt/_modules/metalk8s_volumes.py [hunk 4]
```diff
@@ -132,6 +134,42 @@ def format(name):
...
+        name (str): volume name
+
+    Returns:
+        None
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.clean_up example-volume
```
Comment (NicolasT) [inline salt/_modules/metalk8s_volumes.py]: Same.

File: salt/_modules/metalk8s_volumes.py [hunk 5]
```diff
@@ -260,6 +308,25 @@ def format(self, force=False):
         # We format a "normal" file, not a block device: we need force=True.
         super(SparseLoopDevice, self).format(force=True)
 
+    @property
+    def is_cleaned_up(self):
+        return not (self.is_provisioned or self.exists)
+
+    def clean_up(self):
+        LOOP_CLR_FD = 0x4C01  # From /usr/include/linux/loop.h
```
Comment (NicolasT) [inline salt/_modules/metalk8s_volumes.py:304]: Is this portable across archs?

File: salt/_modules/metalk8s_volumes.py [hunk 6]
```diff
@@ -260,6 +308,25 @@ def format(self, force=False):
...
+        device_path = '/dev/disk/by-uuid/{}'.format(self.get('metadata.uid'))
+        try:
+            with _open_fd(device_path, os.O_RDONLY) as fd:
+                try:
+                    fcntl.ioctl(fd, LOOP_CLR_FD, 0)
+                except Exception:
+                    raise
```
Comment (NicolasT) [inline salt/_modules/metalk8s_volumes.py]: So, no need to `try`/`except`

File: storage-operator/pkg/salt/client.go [hunk 7]
```diff
@@ -95,25 +95,35 @@ func (self *Client) PrepareVolume(
...
 //     ctx:      the request context (used for cancellation)
 //     nodeName: name of the node where the volume will be
+//     volumeName: name of the volume to prepare
+//     saltenv:    saltenv to use
 //
 // Returns
 //     The Salt job ID.
 func (self *Client) UnprepareVolume(
-	ctx context.Context, nodeName string,
+	ctx context.Context, nodeName string, volumeName string, saltenv string,
```
Comment (NicolasT) [inline storage-operator/pkg/salt/client.go:105]: I'm surprised `go fmt` or `go vet` doesn't turn this into `ctx context.Context, nodeName, volumeName, saltenv string`

File: storage-operator/pkg/salt/client.go [hunk 8]
```diff
@@ -95,25 +95,35 @@ func (self *Client) PrepareVolume(
...
+	self.logger.Info(
+		"UnprepareVolume",
+		"Volume.NodeName", nodeName, "Volume.Name", volumeName,
+	)
 
 	ans, err := self.authenticatedRequest(ctx, "POST", "/", payload)
 	if err != nil {
 		return "", errors.Wrapf(
-			err, "UnprepareVolume failed (target=%s)", nodeName,
+			err, "UnrepareVolume failed (target=%s, volume=%s)",
```
Comment (NicolasT) [inline storage-operator/pkg/salt/client.go]: Maybe also log the `saltenv`.

**Review by NicolasT — CHANGES_REQUESTED** [review]

File: salt/_modules/metalk8s_volumes.py (hunk 2 above)
Comment (slaperche-scality) [reply salt/_modules/metalk8s_volumes.py]: Done

File: salt/_modules/metalk8s_volumes.py (hunk 3 above)
Comment (slaperche-scality) [reply salt/_modules/metalk8s_volumes.py]: Indeed. Done

File: salt/_modules/metalk8s_volumes.py (hunk 4 above)
Comment (slaperche-scality) [reply salt/_modules/metalk8s_volumes.py]: Done

File: salt/_modules/metalk8s_volumes.py (hunk 5 above)
Comment (slaperche-scality) [reply salt/_modules/metalk8s_volumes.py:304]: arch as in `endianness`?

I think so, there is no `#define` around the constant in the header.

File: salt/_modules/metalk8s_volumes.py (hunk 6 above)
Comment (slaperche-scality) [reply salt/_modules/metalk8s_volumes.py]: Indeed, it's a vestigial structure.

File: storage-operator/pkg/salt/client.go (hunk 7 above)
Comment (slaperche-scality) [reply storage-operator/pkg/salt/client.go:105]: Indeed, maybe it's not a good practice or not good enough to be enforced.

File: storage-operator/pkg/salt/client.go (hunk 8 above)
Comment (slaperche-scality) [reply storage-operator/pkg/salt/client.go]: Yeah, could be useful.

Done

File: salt/_modules/metalk8s_volumes.py (hunk 5 above)
Comment (slaperche-scality) [reply salt/_modules/metalk8s_volumes.py:304]: I've just tried on big endian (Debian on MIPS) and it works:

```
root@debian:~# truncate -s 1G foo
root@debian:~# losetup -f foo
[   79.088643] loop: module loaded
root@debian:~# losetup -l
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE DIO LOG-SEC
/dev/loop0         0      0         0  0 /root/foo   0     512
root@debian:~# python2
Python 2.7.16 (default, Apr  6 2019, 01:42:57)
[GCC 8.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> platform.machine()
'mips'
>>> import sys
>>> sys.byteorder
'big'
>>> import os
>>> import fcntl
>>> fd = os.open('/dev/loop0', os.O_RDONLY)
>>> fcntl.ioctl(fd, 0x4C01, 0)
0
>>> os.close(fd)
>>>
root@debian:~# losetup -l
root@debian:~#
```

Comment (NicolasT) [reply salt/_modules/metalk8s_volumes.py:304]: :100:

**Review by jbertran — APPROVED** [review]
LGTM

**Review by NicolasT — APPROVED** [review]

**Comment by NicolasT** [comment]
I scheduled a rebuild.

**Pull request merged**